functional-tests/test-bed/src/main/java/org/arquillian/smart/testing/ftest/testbed/project/ProjectBuilder.java

```diff
@@ -113,9 +115,23 @@ private String printSystemProperties() {
     }
 
     private void setCustomMavenInstallation(PomEquippedEmbeddedMaven embeddedMaven) {
-        final String mvnInstallation = System.getenv("TEST_BED_M2_HOME");
-        if (mvnInstallation != null) {
-            embeddedMaven.useInstallation(new File(mvnInstallation));
+        if (buildConfigurator.getMaven3Version() != null && !buildConfigurator.getMaven3Version().isEmpty()) {
+            setMaven3Version(embeddedMaven);
+        } else {
+            final String mvnInstallation = System.getenv("TEST_BED_M2_HOME");
+            if (mvnInstallation != null) {
+                embeddedMaven.useInstallation(new File(mvnInstallation));
+            }
+        }
+    }
+
+    private void setMaven3Version(PomEquippedEmbeddedMaven embeddedMaven) {
+        String maven3Version = buildConfigurator.getMaven3Version();
+        embeddedMaven.useMaven3Version(maven3Version);
+        if (!new File(MavenExtensionRegisterer.DOT_MVN_DIR, MavenExtensionRegisterer.EXTENSIONS_XML).exists()) {
```

Comment:
It would be easier for me to understand what is the condition if this could be extracted to the separated method (or variable) with the self-explaining name. Do you feel like it would improve it too from your perception?